### PR TITLE
Graph full refresh

### PR DESCRIPTION
**Pull request by pkliczewski**
This PR updates v4 spec to verify whether graph refresh creates the same
db entities.

RFE link: https://bugzilla.redhat.com/1466172

### CODE REVIEW

**Comment by pkliczewski** [comment]
This PR depends on https://github.com/ManageIQ/manageiq/pull/15441

**Comment by pkliczewski** [comment]
[Graph refresh](https://trello.com/c/uid9QN8b/161-graph-refresh)

**Comment by pkliczewski** [comment]
This PR requires https://github.com/ManageIQ/manageiq/pull/15441 to work

**Comment by pkliczewski** [comment]
@borod108 @masayag please review

**Comment by pkliczewski** [comment]
@miq-bot assign @oourfali

**Comment by oourfali** [comment]
@miq-bot assign @borod108

Please assign me only once it is approved.

File: app/models/manageiq/providers/redhat/inventory/collector.rb [hunk 1]
```diff
@@ -53,6 +53,24 @@ def collect_datacenters
...
 
+  def collect_cluster_for_host(host)
+    manager.with_provider_connection(VERSION_HASH) do |connection|
+      connection.follow_link(host.cluster)
+    end
+  end
+
+  def collect_host_stats(host)
+    manager.with_provider_connection(VERSION_HASH) do |connection|
+      connection.link?(host.statistics) ? connection.follow_link(host.statistics) : host.statistics
```
Comment (borod108) [inline app/models/manageiq/providers/redhat/inventory/collector.rb]: what is host.statistics if its not link?

File: app/models/manageiq/providers/redhat/inventory/parser/infra_manager.rb [hunk 2]
```diff
@@ -306,70 +357,75 @@ def vms
...
       operating_systems(persister_vm, vm)
       custom_attributes(persister_vm, vm)
     end
   end
 
   def storages(vm)
     storages = []
+    disks = []
     collector.collect_attached_disks(vm).to_miq_a.each do |disk|
+      next if disk.kind_of?(Array) && [].empty?
```
Comment (borod108) [inline app/models/manageiq/providers/redhat/inventory/parser/infra_manager.rb]: [].empty? is always true maybe should be disk.empty?

File: app/models/manageiq/providers/redhat/inventory/parser/infra_manager.rb [hunk 3]
```diff
@@ -306,70 +357,75 @@ def vms
...
-    addresses = hardware_networks(persister_hardware, vm)
+    hardware_disks(persister_hardware, disks)
+    addresses = hardware_networks(persister_hardware, vm, template)
     hardware_guest_devices(persister_hardware, vm, addresses)
   end
 
-  def hardware_networks(persister_hardware, vm)
+  def hardware_networks(persister_hardware, vm, template)
+    addresses = []
+    return addresses if template
```
Comment (borod108) [inline app/models/manageiq/providers/redhat/inventory/parser/infra_manager.rb]: maybe write "unless template" at line 398

File: app/models/manageiq/providers/redhat/inventory/parser/infra_manager.rb [hunk 4]
```diff
@@ -306,70 +357,75 @@ def vms
...
+
     devices = collector.collect_vm_devices(vm)
-    device = devices[0]
+    devices = devices.reject(&:blank?) unless devices.nil?
+    return if devices.blank?
 
+    device = devices[0]
```
Comment (borod108) [inline app/models/manageiq/providers/redhat/inventory/parser/infra_manager.rb]: if this is an array the next line will fail, maybe flatten should be used in line 407?

File: app/models/manageiq/providers/redhat/inventory/parser/infra_manager.rb [hunk 5]
```diff
@@ -306,70 +357,75 @@ def vms
...
 
-  def hardware_networks(persister_hardware, vm)
+  def hardware_networks(persister_hardware, vm, template)
+    addresses = []
+    return addresses if template
+
     devices = collector.collect_vm_devices(vm)
-    device = devices[0]
+    devices = devices.reject(&:blank?) unless devices.nil?
+    return if devices.blank?
```
Comment (borod108) [inline app/models/manageiq/providers/redhat/inventory/parser/infra_manager.rb]: maybe Array(devices).detect{|e| e.present }

File: app/models/manageiq/providers/redhat/inventory/parser/infra_manager.rb [hunk 6]
```diff
@@ -422,9 +483,13 @@ def hardware_guest_devices(persister_hardware, vm, addresses)
   end
 
   def snapshots(vm)
+    snaps = []
+    return snaps if vm.try(:snapshots).nil?
+
     snapshots = collector.collect_snapshots(vm)
+    snapshots = snapshots.sort_by(&:date).reverse!
```
Comment (borod108) [inline app/models/manageiq/providers/redhat/inventory/parser/infra_manager.rb]: I think you can remove the bang

File: app/models/manageiq/providers/redhat/inventory/collector.rb (hunk 1 above)
Comment (pkliczewski) [inline app/models/manageiq/providers/redhat/inventory/collector.rb]: It looks like host.statistics can be nil in situation like host being installed so there are no stats yet.

**Comment by pkliczewski** [comment]
@miq-bot assign @oourfali

**Pull request merged**